packages/harden/src/make-hardener.js

```diff
@@ -0,0 +1,285 @@
+// Adapted from SES/Caja - Copyright (C) 2011 Google Inc.
+// Copyright (C) 2018 Agoric
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// based upon:
+// https://github.com/google/caja/blob/master/src/com/google/caja/ses/startSES.js
+// https://github.com/google/caja/blob/master/src/com/google/caja/ses/repairES5.js
+// then copied from proposal-frozen-realms deep-freeze.js
+// then copied from SES/src/bundle/deepFreeze.js
+
+// @ts-check
+
+import {
+  Set,
+  String,
+  TypeError,
+  WeakSet,
+  globalThis,
+  apply,
+  arrayForEach,
+  defineProperty,
+  freeze,
+  getOwnPropertyDescriptor,
+  getOwnPropertyDescriptors,
+  getPrototypeOf,
+  isInteger,
+  isObject,
+  objectHasOwnProperty,
+  ownKeys,
+  preventExtensions,
+  setAdd,
+  setForEach,
+  setHas,
+  toStringTagSymbol,
+  typedArrayPrototype,
+  weaksetAdd,
+  weaksetHas,
+  isError,
+} from '@endo/intrinsics';
+import { FERAL_STACK_GETTER, FERAL_STACK_SETTER } from './commons.js';
+
+const assert = guard => {
+  if (!guard) {
+    throw new TypeError('assertion failed');
+  }
+};
+
+/**
+ * @import {Harden} from '../types.js'
+ */
+
+// Obtain the string tag accessor of of TypedArray so we can indirectly use the
+// TypedArray brand check it employs.
+const typedArrayToStringTag = getOwnPropertyDescriptor(
```

Comment:
Separately, I have been thinking that once Immutable ArrayBuffers are in the language and adequately deployed, we should retire harden's special case for TypedArrays. If someone wants to harden a TypedArray, they should start with a frozen TypedArray.